url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/factory/FCS.java

```diff
@@ -0,0 +1,248 @@
+		queue = q;
+		updateQueue();
+	}
+
+	public boolean newPart(PartType pt) {
+		for (PartType p : partTypes) {
+			if (p.getName().equals(pt.getName())) {
+				displayMessage("Professor Oak: Professor Oak: Part \"" + pt.getName() + "\" already exists!");
+				return false;
+			}
+		}
+
+		partTypes.add(pt);
+		updateParts();
+		agent.msgAddNewPartType(pt);
+
+		displayMessage("Professor Oak: Part added - " + pt.getName());
+		return true;
+	}
+
+	public void editPart(PartType pt) {
+		// loops through ArrayList to find based on id, then replaces
+		for (int i = 0; i < partTypes.size(); i++) {
+			if (partTypes.get(i).equals(pt)) {
+				partTypes.set(i, pt);
+				updateParts();
+				GantryGraphics gantry = (GantryGraphics) server.devices.get(Constants.GANTRY_ROBOT_TARGET);
+				gantry.editBin(pt);
+			}
+		}
+
+		displayMessage("Professor Oak: Part edited - " + pt.getName());
+	}
+
+	public void deletePart(PartType pt) {
+		// loops through ArrayList to find based on id, then deletes
+		for (int i = 0; i < partTypes.size(); i++) {
+			if (partTypes.get(i).equals(pt)) {
+				partTypes.remove(i);
+				updateParts();
+			}
+		}
+
+		displayMessage("Professor Oak: Part deleted - " + pt.getName());
+		// TODO: add in agent call so to stop drawing bin
+	}
+
+	public boolean newKit(KitConfig kc) {
+		for (KitConfig p : kitConfigs) {
+			if (p.getName().equals(kc.getName())) {
+				displayMessage("Professor Oak: Professor Oak: Kit \"" + kc.getName() + "\" already exists!");
+				return false;
+			}
+		}
+		kitConfigs.add(kc);
+		updateKits();
+
+		displayMessage("Professor Oak: Kit added - " + kc.getName());
+		return true;
+	}
+
+	public void editKit(KitConfig kc) {
+		// loops through ArrayList to find based on id, then replaces
+		for (int i = 0; i < kitConfigs.size(); i++) {
+			if (kitConfigs.get(i).equals(kc)) {
+				kitConfigs.set(i, kc);
+			}
+		}
+		updateKits();
+		displayMessage("Professor Oak: Kit edited - " + kc.getName());
+	}
+
+	public void deleteKit(KitConfig kc) {
+		// loops through ArrayList to find based on id, then deletes
+		for (int i = 0; i < kitConfigs.size(); i++) {
+			if (kitConfigs.get(i).equals(kc)) {
+				kitConfigs.remove(i);
+			}
+		}
+		updateKits();
+		displayMessage("Professor Oak: Kit deleted - " + kc.getName());
+	}
+
+	public void addOrder(Order o) {
+		if (o.getNumKits() > 0) {
+			agent.msgAddKitsToQueue(o);
+
+			// TODO: put "start production" button
+			if (!productionStarted) {
+				agent.msgStartProduction();
+				productionStarted = true;
+			}
+		}
+		displayMessage("Professor Oak: Ordered " + o.getConfig().getName() + " x" + o.getNumKits());
+	}
+
+	public void startProduction() {
+		agent.msgStartProduction();
+	}
+
+	public void setDropChance(Float c) {
+		// TODO Make a slider/button for this
+		agent.msgSetPartsRobotDropChance(c);
+		displayMessage("Professor Oak: Parts Robot's drop rate set to " + c * 100 + "%");
+	}
+
+	public void receiveData(Request req) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.